wagtail/wagtailcore/models.py

```diff
@@ -625,12 +625,29 @@ def is_navigable(self):
         """
         return (not self.is_leaf()) or self.depth == 2
 
+    def get_url_parts(self):
+        """
+        Determine the URL for this page and return it as a tuple of
+        (site_id, site_root_url, page_url_relative_to_site_root).
+        Return None if the page is not routable.
+
+        Pages with custom URL routing should override this.
+        """
+        for (site_id, root_path, root_url) in Site.get_site_root_paths():
+            if self.url_path.startswith(root_path):
+                page_path = reverse('wagtail_serve', args=(self.url_path[len(root_path):],))
+                return (site_id, root_url, page_path)
+
     @property
     def full_url(self):
         """Return the full URL (including protocol / domain) to this page, or None if it is not routable"""
-        for (id, root_path, root_url) in Site.get_site_root_paths():
-            if self.url_path.startswith(root_path):
-                return root_url + reverse('wagtail_serve', args=(self.url_path[len(root_path):],))
+        try:
+            site_id, root_url, page_path = self.get_url_parts()
+        except TypeError:
```

Comment:
Good point! I was trying to follow the Python 'EAFP' principle, but that often has the risk of catching more exceptions than you intended, as it does here.